From 5f5c6d7d1070a4a7d248e2f92e480f55ada895cc Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Wed, 4 Jun 2025 12:33:56 +0200
Subject: [PATCH 0961/1010] evl/net: skb: enable page pool for pseudo-devices

We may enable the oob mode for pseudo-devices like the loopback
interface (lo). Make sure to skip the DMA settings when creating the
page pool for these.

At this chance, also refrain from dereferencing dev.parent which might
be NULL when determining the NUMA node.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 kernel/evl/net/skb.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/kernel/evl/net/skb.c b/kernel/evl/net/skb.c
index fdac81837b..4a36e6c4aa 100644
--- a/kernel/evl/net/skb.c
+++ b/kernel/evl/net/skb.c
@@ -521,7 +521,7 @@ int evl_net_dev_build_pool(struct net_device *dev)
 		.order = ilog2(est->buf_size / PAGE_SIZE),
 		.flags = PP_FLAG_PAGE_OOB,
 		.pool_size = est->pool_max,
-		.nid = dev_to_node(dev->dev.parent),
+		.nid = dev->dev.parent ? dev_to_node(dev->dev.parent) : NUMA_NO_NODE,
 		.dev = dev->dev.parent,
 		.dma_dir = DMA_NONE,
 		.offset = 0,
@@ -529,13 +529,14 @@ int evl_net_dev_build_pool(struct net_device *dev)
 	};
 
 	/*
-	 * If the device is oob-capable, the page pool must perform
-	 * DMA pre-mapping so that the NIC driver only has to deal
-	 * with cache synchronization on the out-of-band TX path.  We
-	 * are piggybacked by the XDP/TX support which enables
-	 * DMA_BIDIRECTIONAL (DMA_TO_DEVICE is not supported).
+	 * If the netdev is an oob-capable physical device, the page
+	 * pool must perform DMA pre-mapping so that the NIC driver
+	 * only has to deal with cache synchronization on the
+	 * out-of-band TX path.  We are piggybacked by the XDP/TX
+	 * support which enables DMA_BIDIRECTIONAL (DMA_TO_DEVICE is
+	 * not supported).
 	 */
-	if (netdev_is_oob_capable(dev)) {
+	if (pp_params.dev && netdev_is_oob_capable(dev)) {
 		pp_params.flags |= PP_FLAG_DMA_MAP | PP_FLAG_DMA_SYNC_DEV;
 		pp_params.dma_dir = DMA_BIDIRECTIONAL;
 	}
-- 
2.30.2

